anvill/src/Lifters/FunctionLifter.cpp

```diff
@@ -0,0 +1,1389 @@
+#include <anvill/TypePrinter.h>
+
+#include <remill/Arch/Arch.h>
+#include <remill/Arch/Instruction.h>
+#include <remill/BC/Compat/Error.h>
+#include <remill/BC/Util.h>
+#include <remill/BC/Version.h>
+#include <remill/OS/OS.h>
+
+#include <llvm/IR/BasicBlock.h>
+#include <llvm/IR/DerivedTypes.h>
+#include <llvm/IR/Function.h>
+#include <llvm/IR/Instruction.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/Intrinsics.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/LegacyPassManager.h>
+#include <llvm/IR/Module.h>
+#include <llvm/IR/Type.h>
+
+#include <llvm/Transforms/Scalar.h>
+#include <llvm/Transforms/Utils.h>
+#include <llvm/Transforms/Utils/Cloning.h>
+#include <llvm/Transforms/InstCombine/InstCombine.h>
+
+#include <sstream>
+
+#include "Context.h"
+
+// TODO(pag): Externalize this into some kind of `LifterOptions` struct.
+DEFINE_bool(print_registers_before_instuctions, false,
+            "Inject calls to printf (into the lifted bitcode) to log integer "
+            "register state to stdout.");
+
+namespace anvill {
+namespace {
+
+// Clear out LLVM variable names. They're usually not helpful.
+static void ClearVariableNames(llvm::Function *func) {
+  for (auto &block : *func) {
+    block.setName(llvm::Twine::createNull());
+    for (auto &inst : block) {
+      if (inst.hasName()) {
+        inst.setName(llvm::Twine::createNull());
+      }
+    }
+  }
+}
+
+// Compatibility function for performing a single step of inlining.
+static llvm::InlineResult InlineFunction(llvm::CallBase *call,
+                                         llvm::InlineFunctionInfo &info) {
+#if LLVM_VERSION_NUMBER < LLVM_VERSION(11, 0)
+  return llvm::InlineFunction(call, info);
+#else
+  return llvm::InlineFunction(*call, info);
+#endif
+}
+
+// A function that ensures that the memory pointer escapes, and thus none of
+// the memory writes at the end of a function are lost.
+static llvm::Function *
+GetMemoryEscapeFunc(const remill::IntrinsicTable &intrinsics) {
+  const auto module = intrinsics.error->getParent();
+  auto &context = module->getContext();
+
+  const auto name = "__anvill_memory_escape";
+  if (auto func = module->getFunction(name)) {
+    return func;
+  }
+
+  llvm::Type *params[] = {
+      remill::NthArgument(intrinsics.error, remill::kMemoryPointerArgNum)
+          ->getType()};
+  auto type =
+      llvm::FunctionType::get(llvm::Type::getVoidTy(context), params, false);
+  return llvm::Function::Create(type, llvm::GlobalValue::ExternalLinkage, name,
+                                module);
+}
+
+}  // namespace
+
+FunctionLifterImpl::~FunctionLifterImpl(void) {}
+
+FunctionLifterImpl::FunctionLifterImpl(
+    const LifterOptions &options_, MemoryProvider &memory_provider_,
+    TypeProvider &type_provider_)
+    : options(options_),
+      memory_provider(memory_provider_),
+      type_provider(type_provider_),
+      semantics_module(remill::LoadArchSemantics(options.arch)),
+      context(semantics_module->getContext()),
```

Comment:
Might be a good idea to rename this to reflect that this is an `LLVMContext`.

Reply:
Renamed to `llvm_context`